### PR TITLE
Add com.gluonhq.SceneBuilder

### PR DESCRIPTION
**Pull request by ghazyami**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link: https://github.com/gluonhq/scenebuilder/issues/293**
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by ghazyami** [comment]
There is an ongoing effort to submit this before, I tried contacting the submitter but could not get a response so far here
https://github.com/flathub/flathub/pull/3402

**Comment by ghazyami** [comment]
bot, build com.gluonhq.SceneBuilder

**Comment by ghazyami** [comment]
bot, build com.gluonhq.SceneBuilder

**Comment by hfiguiere** [comment]
There is already PR #3402 for this app.

File: com.gluonhq.SceneBuilder.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
+app-id: com.gluonhq.SceneBuilder
+runtime: org.freedesktop.Platform
+runtime-version: "22.08"
+sdk: org.freedesktop.Sdk
+command: SceneBuilder-runner
+rename-desktop-file: scenebuilder-SceneBuilder.desktop
+finish-args:
+  - --env=PATH=/app/bin
```
Comment (hfiguiere) [inline com.gluonhq.SceneBuilder.yml:8]:
```suggestion
```

File: com.gluonhq.SceneBuilder.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  - --socket=x11
+  - --share=ipc
+  - --share=network
+  - --persist=.scenebuilder
+  - --filesystem=home
+cleanup:
+  - /share/info/*
+  - /share/locale/*
+  - /share/man/*
+  - /share/runtime/locale/*
```
Comment (hfiguiere) [inline com.gluonhq.SceneBuilder.yml:18]: why do you delete the locales?

File: com.gluonhq.SceneBuilder.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+modules:
+  - name: SceneBuilder
+    buildsystem: simple
+    build-commands:
+    - mkdir -p deb-package
+    - ar p SceneBuilder.deb data.tar.xz | tar -xJf - -C deb-package
+    - mv deb-package/opt/scenebuilder /app
+    - install -dm755 /app/bin
+    - install -m755 SceneBuilder-runner /app/bin/SceneBuilder-runner
```
Comment (hfiguiere) [inline com.gluonhq.SceneBuilder.yml:27]:
```suggestion
    - install -Dm755 SceneBuilder-runner /app/bin/SceneBuilder-runner
```

File: com.gluonhq.SceneBuilder.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    - install -Dm644 com.gluonhq.SceneBuilder.metainfo.xml -t /app/share/metainfo/
+    - install -Dm644 com.gluonhq.SceneBuilder.256.png /app/share/icons/hicolor/256x256/apps/com.gluonhq.SceneBuilder.png
+    - install -Dm644 com.gluonhq.SceneBuilder.128.png /app/share/icons/hicolor/128x128/apps/com.gluonhq.SceneBuilder.png
+    - desktop-file-edit --remove-key=Categories /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-edit --set-key=Categories --set-value='Development' /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-edit --set-icon=com.gluonhq.scenebuilder /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-edit --remove-key=Exec /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-edit --set-key=Exec --set-value='SceneBuilder-runner' /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
```
Comment (hfiguiere) [inline com.gluonhq.SceneBuilder.yml:35]: Afaik you don't need to remove it.
```suggestion
    - desktop-file-edit --set-key=Exec --set-value='SceneBuilder-runner' /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
```

File: com.gluonhq.SceneBuilder.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    - mkdir -p deb-package
+    - ar p SceneBuilder.deb data.tar.xz | tar -xJf - -C deb-package
+    - mv deb-package/opt/scenebuilder /app
+    - install -dm755 /app/bin
+    - install -m755 SceneBuilder-runner /app/bin/SceneBuilder-runner
+    - install -Dm644 com.gluonhq.SceneBuilder.metainfo.xml -t /app/share/metainfo/
+    - install -Dm644 com.gluonhq.SceneBuilder.256.png /app/share/icons/hicolor/256x256/apps/com.gluonhq.SceneBuilder.png
+    - install -Dm644 com.gluonhq.SceneBuilder.128.png /app/share/icons/hicolor/128x128/apps/com.gluonhq.SceneBuilder.png
+    - desktop-file-edit --remove-key=Categories /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-edit --set-key=Categories --set-value='Development' /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
```
Comment (hfiguiere) [inline com.gluonhq.SceneBuilder.yml:32]: why no using `add-category` ?

File: com.gluonhq.SceneBuilder.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    - desktop-file-edit --set-icon=com.gluonhq.scenebuilder /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-edit --remove-key=Exec /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-edit --set-key=Exec --set-value='SceneBuilder-runner' /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-install /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop --dir=/app/share/applications
+
+    sources:
+    - type: file
+      dest-filename: SceneBuilder.deb
+      url: https://github.com/gluonhq/scenebuilder/releases/download/18.0.0/SceneBuilder-18.0.0.deb
```
Comment (hfiguiere) [inline com.gluonhq.SceneBuilder.yml:41]: so you are building an aarch64 flatpak using what I assume is x86_64 only binaries.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: com.gluonhq.SceneBuilder.yml (hunk 2 above)
Comment (ghazyami) [reply com.gluonhq.SceneBuilder.yml:17]: I wasn't sure about this, took it from the original PR #3402
Will remove them now

File: com.gluonhq.SceneBuilder.yml (hunk 4 above)
Comment (ghazyami) [reply com.gluonhq.SceneBuilder.yml:33]: I removed it because I got an error on my local machine about the value

Comment (hfiguiere) [reply com.gluonhq.SceneBuilder.yml:33]: set ought be be enough.

File: com.gluonhq.SceneBuilder.yml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    - mkdir -p deb-package
+    - ar p SceneBuilder.deb data.tar.xz | tar -xJf - -C deb-package
+    - mv deb-package/opt/scenebuilder /app
+    - install -Dm755 SceneBuilder-runner /app/bin/SceneBuilder-runner
+    - install -Dm644 com.gluonhq.SceneBuilder.metainfo.xml -t /app/share/metainfo/
+    - install -Dm644 com.gluonhq.SceneBuilder.256.png /app/share/icons/hicolor/256x256/apps/com.gluonhq.SceneBuilder.png
+    - install -Dm644 com.gluonhq.SceneBuilder.128.png /app/share/icons/hicolor/128x128/apps/com.gluonhq.SceneBuilder.png
+    - desktop-file-edit --remove-key=Categories /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-edit --set-key=Categories --set-value='Development' /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
+    - desktop-file-edit --set-icon=com.gluonhq.scenebuilder /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
```
Comment (hfiguiere) [inline com.gluonhq.SceneBuilder.yml:31]: icon name are case sensitive like file names.
```suggestion
    - desktop-file-edit --set-icon=com.gluonhq.SceneBuilder /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
```

File: com.gluonhq.SceneBuilder.yml (hunk 6 above)
Comment (ghazyami) [reply com.gluonhq.SceneBuilder.yml:37]: Added `build-options` to run only on x86_64

File: com.gluonhq.SceneBuilder.yml (hunk 7 above)
Comment (ghazyami) [reply com.gluonhq.SceneBuilder.yml]: already updated it

File: com.gluonhq.SceneBuilder.yml (hunk 5 above)
Comment (ghazyami) [reply com.gluonhq.SceneBuilder.yml:28]: `Running: desktop-file-edit --set-icon=com.gluonhq.scenebuilder /app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop
/app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop: error: value ""Development;Building;GUIDesigner;Java;"" for key "Categories" in group "Desktop Entry" contains an unregistered value ""Development"; values extending the format should start with "X-"
/app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop: error: value ""Development;Building;GUIDesigner;Java;"" for key "Categories" in group "Desktop Entry" contains an unregistered value """; values extending the format should start with "X-"
Error on file "/app/scenebuilder/lib/scenebuilder-SceneBuilder.desktop": Failed to validate the created desktop file
`

Comment (ghazyami) [reply com.gluonhq.SceneBuilder.yml:28]: original value seem to be causing issues

**Comment by ghazyami** [comment]
bot, build com.gluonhq.SceneBuilder

File: com.gluonhq.SceneBuilder.yml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+rename-desktop-file: scenebuilder-SceneBuilder.desktop
+finish-args:
+  - --socket=x11
+  - --share=ipc
+  - --share=network
+  - --persist=.scenebuilder
+  - --filesystem=home
+build-options:
+  arch:
+    x86_64:
```
Comment (hfiguiere) [inline com.gluonhq.SceneBuilder.yml:15]: this has strictly no effect.

Sicen  you copied #3402, there is a hint over there.

**Comment by hfiguiere** [comment]
I wish this was built from source to support aarch64 as well.

**Comment by ghazyami** [comment]
Okay, will try to build from source

**Comment by ghazyami** [comment]
@hfiguiere 
Okay, since this is a Java application building from source requires Java to be available, does `org.freedesktop.Sdk.Extension.openjdk` extension, have a `aarch64`  version?
We also need `maven` packaging tool (downloaded as regular file? Or there is a `buildsystem`).

File: com.gluonhq.SceneBuilder.yml (hunk 8 above)
Comment (hfiguiere) [inline com.gluonhq.SceneBuilder.yml:15]:
```suggestion
```

**Comment by JakobDev** [comment]
> I wish this was built from source to support aarch64 as well.

It is Java. It works on all Acrhitectures without buidling from source.

>  Okay, since this is a Java application building from source requires Java to be available, does org.freedesktop.Sdk.Extension.openjdk extension, have a aarch64 version?

Yes

>  We also need maven packaging tool (downloaded as regular file? Or there is a buildsystem).

There is currently nothing.

I would suggest that you don't build from source. Just install the JAR file and use the Java runtime from `org.freedesktop.Sdk.Extension.openjdk` instead of the one included in the Deb and it should work on ARM too.

**Comment by ghazyami** [comment]
@JakobDev unfortunately, they don't provide a `.jar` artifact, the only thing they provide as a jar provide something that is called `Scene Builder Kit` which used to embed scene builder into other products,

I think we can only produce the jar from source or take out of the `.deb` for example,
Apart from regular Java runtime provided by `org.freedesktop.Sdk.Extension.openjdk` we also need to download `JavaFX` SDK which is arch specific that could be solved via (files on the sources section I guess)

**Comment by ghazyami** [comment]
> @JakobDev unfortunately, they don't provide a `.jar` artifact, the only thing they provide as a jar is called `Scene Builder Kit` which used to embed scene builder into other products,
> 
> I think we can only produce the jar from source or take out of the `.deb` for example, Apart from regular Java runtime provided by `org.freedesktop.Sdk.Extension.openjdk` we also need to download `JavaFX` SDK which is arch specific that could be solved via (files on the sources section I guess)

The other option is continuing the current approach, because the `.deb` contains a subset of the Java runtime (the parts only needed by scene builder + the JavaFX SDK)

**Comment by batashi** [comment]
Great efforts  @ghazyami 
sorry was busy in the last days.

**Pull request closed without merging**